doc/source/admin/install/config_kubernetes.rst
..
      Copyright 2018 Catalyst IT Ltd
      All Rights Reserved.
      not use this file except in compliance with the License. You may obtain
      a copy of the License at

          http://www.apache.org/licenses/LICENSE-2.0

      Unless required by applicable law or agreed to in writing, software
      distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
      WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
      License for the specific language governing permissions and limitations
      under the License.

Config Qinling with existing kubernetes/etcd cluster
====================================================

In most cases, it's not ideal to set up a new dedicated kubernetes cluster for
Qinling. The component which works with kubernetes cluster in Qinling is the
``qinling-engine``. Follow the steps below to configure Qinling to work with
existing kubernetes/etcd cluster, and make Qinling access the kubernetes/etcd
service with authentication and authorization.

Prerequisites
~~~~~~~~~~~~~

* You know the kubernetes API address and etcd service address, for example:

  .. code-block:: console

      export K8S_ADDRESS=10.0.0.5
      export ETCD_ADDRESS=10.0.0.6
      export QINLING_SERVICE_USER=qinling

  .. end

  Make sure the kubernetes and etcd services are both accessible to external.

* You know the IP address that ``qinling-engine`` service talks to kubernetes,
  for example:

  .. code-block:: console

      export QINLING_ENGINE_ADDRESS=10.0.0.7

  .. end

* You have CA certificates of the kubernetes and etcd respectively and store on
  the host that ``qinling-engine`` is running.

  .. code-block:: console

      export K8S_CA_CERT=$HOME/ca.crt
      export K8S_CA_KEY=$HOME/ca.key
      export ETCD_CA_CERT=$HOME/etcd_ca.crt
      export ETCD_CA_KEY=$HOME/etcd_ca.key

  .. end

* This guide assumes
  `RBAC <https://kubernetes.io/docs/admin/authorization/rbac/>`_ is enabled in
  the kubernetes cluster.
* All the following commands are supposed to be executed under ``root``
  permission.

Qinling configurations
~~~~~~~~~~~~~~~~~~~~~~

Below are the options and their default values that relate to accessing the
Kubernetes API and etcd in Qinling's configuration file.

.. code-block:: ini

    [kubernetes]
    kube_host = https://127.0.0.1:8001
    use_api_certificate = True
    ssl_ca_cert = /etc/qinling/pki/kubernetes/ca.crt
    cert_file = /etc/qinling/pki/kubernetes/qinling.crt
    key_file = /etc/qinling/pki/kubernetes/qinling.key

    [etcd]
    host = 127.0.0.1
    port = 2379
    protocol = https
    ca_cert = /etc/qinling/pki/etcd/ca.crt
    cert_file = /etc/qinling/pki/etcd/qinling-etcd-client.crt
    cert_key = /etc/qinling/pki/etcd/qinling-etcd-client.key

.. end

First, change the kubernetes and etcd service addresses in the config file, and
add the addresses that ``qinling-engine`` uses to talk to kubernetes services
to the ``trusted_cidrs`` option. We will create all the related certificates in
the following steps.

.. note::

    If the ``qinling-engine`` service is running behind a NAT device, make sure
    you get the correct IP address that talks to kubernetes.

.. code-block:: ini

    [kubernetes]
    kube_host = https://${K8S_ADDRESS}:6443
    trusted_cidrs = ${QINLING_ENGINE_ADDRESS}/32
    ...
    [etcd]
    host = ${ETCD_ADDRESS}
    ...

.. end

Generate and config client certificates for Qinling
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

There are a lot of
`tools <https://kubernetes.io/docs/concepts/cluster-administration/certificates/>`_
out there for certificate generation. We use ``cfssl`` as the example here.

#) Download and prepare the command line tools as needed.

    .. code-block:: console

        curl -L https://pkg.cfssl.org/R1.2/cfssl_linux-amd64 -o /tmp/cfssl
        chmod +x /tmp/cfssl
        curl -L https://pkg.cfssl.org/R1.2/cfssljson_linux-amd64 -o /tmp/cfssljson
        chmod +x /tmp/cfssljson

    .. end

#) Generate the kubernetes and etcd client certificates for Qinling.

    .. code-block:: console

        mkdir -p /tmp/certs; cd /tmp/certs
        curl -SL https://raw.githubusercontent.com/openstack/qinling/master/example/kubernetes/cfssl-ca-config.json -o /tmp/certs/cfssl-ca-config.json
        curl -SL https://raw.githubusercontent.com/openstack/qinling/master/example/kubernetes/cfssl-client-csr.json -o /tmp/certs/cfssl-client-csr.json
        /tmp/cfssl gencert -ca=${K8S_CA_CERT} \
            -ca-key=${K8S_CA_KEY} \
            -config=/tmp/certs/cfssl-ca-config.json \
            -profile=client \
            /tmp/certs/cfssl-client-csr.json | /tmp/cfssljson -bare k8s-client
        /tmp/cfssl gencert -ca=${ETCD_CA_CERT} \
            -ca-key=${ETCD_CA_KEY} \
            -config=/tmp/certs/cfssl-ca-config.json \
            -profile=client \
            /tmp/certs/cfssl-client-csr.json | /tmp/cfssljson -bare etcd-client

    .. end

#) Move the certificates to the pre-defined locations in the config file and
   ensure the qinling service user has the permission to those locations.

    .. code-block:: console

        mkdir -p /etc/qinling/pki/{kubernetes,etcd}
        cp k8s-client-key.pem /etc/qinling/pki/kubernetes/qinling.key
        cp k8s-client.pem /etc/qinling/pki/kubernetes/qinling.crt
        cp etcd-client-key.pem /etc/qinling/pki/etcd/qinling-etcd-client.key
        cp etcd-client.pem /etc/qinling/pki/etcd/qinling-etcd-client.crt
        cp ${K8S_CA_CERT} /etc/qinling/pki/kubernetes/ca.crt
        cp ${ETCD_CA_CERT} /etc/qinling/pki/etcd/ca.crt
        chown -R ${QINLING_SERVICE_USER}:${QINLING_SERVICE_USER} /etc/qinling/pki
        cd -; rm -rf /tmp/certs

    .. end

Create Role and RoleBinding in Kubernetes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

According to the least privilege principle, the operation permission of qinling
service user in kubernetes cluster should be restricted, this could be easily
achieved by applying the pre-defined authorization manifest file. The following
command is supposed to be executed with ``admin`` access of the kubernetes
cluster.

.. code-block:: console

    curl -sSL https://raw.githubusercontent.com/openstack/qinling/master/example/kubernetes/k8s_qinling_role.yaml | kubectl apply -f -

.. end

Restart Qinlig services
~~~~~~~~~~~~~~~~~~~~~~~

Restart all the Qinling services. Now Qinling is accessing the Kubernetes API
and etcd service using TLS. The requests that Qinling makes to the Kubernetes
API are also authorized.

.. code-block:: console

    systemctl restart devstack@qinling-*.service

.. end

Access the Kubernetes API Insecurely (For testing purpose ONLY)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Qinling can also connect to the Kubernetes API insecurely if the Kubernetes API
server serves for insecure connections. However, this is not recommended and
should be used for testing purpose only.

In the configuration file, under the ``kubernetes`` section, set ``kube_host``
to the URI which the Kubernetes API serves for insecure HTTP connections, for
example, ``kube_host = http://localhost:8080``, and set ``use_api_certificate``
to ``False`` to disable Qinling using a client certificate to access the
Kubernetes API.

.. code-block:: ini

    [kubernetes]
    kube_host = http://localhost:8080
    use_api_certificate = False

.. end


doc/source/contributor/development-environment-devstack.rst
..
      Copyright 2017 Catalyst IT Ltd
      All Rights Reserved.
      not use this file except in compliance with the License. You may obtain
      a copy of the License at

          http://www.apache.org/licenses/LICENSE-2.0

      Unless required by applicable law or agreed to in writing, software
      distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
      WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
      License for the specific language governing permissions and limitations
      under the License.

Setting up a development environment with devstack
==================================================

This page describes how to set up a working development
environment that can be used in deploying qinling on latest releases
of Ubuntu. These instructions assume you are already familiar
with git. Refer to `Getting the code`_ for additional information.

.. _Getting the code: http://wiki.openstack.org/GettingTheCode

Following these instructions will allow you to have a fully functional qinling
environment using the devstack project (a shell script to build
complete OpenStack development environments) on Ubuntu 16.04.

Configuring devstack with Qinling
---------------------------------

Qinling can be enabled in devstack by using the plug-in based interface it
offers.

.. note::

   The following steps have been fully verified only on Ubuntu 16.04.

Start by cloning the devstack repository using a non-root user:

::

    git clone https://github.com/openstack-dev/devstack

Change to devstack directory:

::

    cd devstack/

Create the ``local.conf`` file with the following minimal devstack
configuration:

.. code-block:: ini

    [[local|localrc]]
    RECLONE=True
    enable_plugin qinling https://github.com/openstack/qinling

    LIBS_FROM_GIT=python-qinlingclient
    DATABASE_PASSWORD=password
    ADMIN_PASSWORD=password
    SERVICE_PASSWORD=password
    SERVICE_TOKEN=password
    RABBIT_PASSWORD=password
    LOGFILE=$DEST/logs/stack.sh.log
    LOG_COLOR=False
    LOGDAYS=1

    ENABLED_SERVICES=rabbit,mysql,key

.. end

Here are several things you could customize:

* For multiple network cards, you need to specify the kubernetes API server's
  advertise address manually.

   .. code-block:: console

       export EXTRA_KUBEADM_INIT_OPTS="--apiserver-advertise-address <default-host-ip>"

   .. end

* Devstack will set up a new kubernetes cluster and re-use etcd service inside
  the cluster for Qinling services, which means you don't need to add etcd to
  the enabled services list in the ``local.conf`` file.
* If you already have an existing kubernetes/etcd cluster, add
  ``QINLING_INSTALL_K8S=False`` to the ``local.conf`` file. You need to
  manually config Qinling services after devstack completes, go to
  `Config Qinling with existing Kubernetes cluster <https://docs.openstack.org/qinling/latest/admin/install/config_kubernetes.html>`_
  for more configuration details.
* If you want to interact with Qinling in Horizon dashboard, add the following
  line to the ``local.conf`` file.

    .. code-block:: console

        enable_plugin qinling-dashboard https://git.openstack.org/openstack/qinling-dashboard

    .. end

Running devstack
----------------

Run the ``stack.sh`` script:

::

    ./stack.sh

After it completes, verify qinling service is installed properly:

.. code-block:: console

    $ source openrc admin admin
    $ openstack service list
    +----------------------------------+----------+-----------------+
    | ID                               | Name     | Type            |
    +----------------------------------+----------+-----------------+
    | 59be2ecc8b8d4e61af184ea3495bf207 | qinling  | function-engine |
    | e5891d41a929402384ef00ce7135a16d | keystone | identity        |
    +----------------------------------+----------+-----------------+
    $ openstack runtime list --print-empty
    +----+------+-------+--------+-------------+------------+------------+------------+
    | Id | Name | Image | Status | Description | Project_id | Created_at | Updated_at |
    +----+------+-------+--------+-------------+------------+------------+------------+
    +----+------+-------+--------+-------------+------------+------------+------------+

.. end

Kubernetes Integration
----------------------

By default, Qinling uses Kubernetes as its orchestrator backend, so a k8s
all-in-one environment (and some other related tools, e.g. kubectl) is also
set up during devstack installation.

Qinling devstack script uses `kubeadm <https://kubernetes.io/docs/setup/independent/create-cluster-kubeadm/>`_
for Kubernetes installation, refer to ``tools/gate/kubeadm/setup_gate.sh`` for
more detailed information about Qinling devstack installation.

qinling/config.py
# Copyright 2017 Catalyst IT Limited
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
from keystoneauth1 import loading
from keystonemiddleware import auth_token
from oslo_concurrency import processutils
from oslo_config import cfg
from oslo_log import log

from qinling import version

CONF = cfg.CONF

launch_opt = cfg.ListOpt(
    'server',
    default=['all'],
    help='Specifies which qinling server to start by the launch script.'
)

default_opts = [
    cfg.StrOpt(
        'qinling_endpoint',
        help='Qinling service endpoint.'
    ),
]

API_GROUP = 'api'
api_opts = [
    cfg.StrOpt('host', default='0.0.0.0', help='Qinling API server host.'),
    cfg.PortOpt('port', default=7070, help='Qinling API server port.'),
    cfg.BoolOpt(
        'enable_ssl_api',
        default=False,
        help='Enable the integrated stand-alone API to service requests'
             'via HTTPS instead of HTTP.'
    ),
    cfg.IntOpt(
        'api_workers',
        default=processutils.get_worker_count(),
        help='Number of workers for Qinling API service '
             'default is equal to the number of CPUs available if that can '
             'be determined, else a default worker count of 1 is returned.'
    ),
    cfg.BoolOpt(
        'enable_job_handler',
        default=True,
        help='Enable job handler.'
    ),
]

PECAN_GROUP = 'pecan'
pecan_opts = [
    cfg.StrOpt(
        'root',
        default='qinling.api.controllers.root.RootController',
        help='Pecan root controller'
    ),
    cfg.ListOpt(
        'modules',
        default=["qinling.api"],
        help='A list of modules where pecan will search for applications.'
    ),
    cfg.BoolOpt(
        'debug',
        default=False,
        help='Enables the ability to display tracebacks in the browser and'
             ' interactively debug during development.'
    ),
    cfg.BoolOpt(
        'auth_enable',
        default=True,
        help='Enables user authentication in pecan.'
    )
]

ENGINE_GROUP = 'engine'
engine_opts = [
    cfg.StrOpt(
        'host',
        default='0.0.0.0',
        help='Name of the engine node. This can be an opaque '
             'identifier. It is not necessarily a hostname, '
             'FQDN, or IP address.'
    ),
    cfg.StrOpt(
        'topic',
        default='qinling_engine',
        help='The message topic that the engine listens on.'
    ),
    cfg.StrOpt(
        'orchestrator',
        default='kubernetes',
        choices=['kubernetes', 'swarm'],
        help='The container orchestrator.'
    ),
    cfg.IntOpt(
        'function_service_expiration',
        default=3600,
        help='Maximum service time in seconds for function in orchestrator.'
    ),
    cfg.IntOpt(
        'function_concurrency',
        default=3,
        help='Maximum number of concurrent executions per function.'
    ),
    cfg.StrOpt(
        'sidecar_image',
        default='openstackqinling/sidecar:0.0.2',
        help='The sidecar image being used together with the worker.'
    ),
]

STORAGE_GROUP = 'storage'
storage_opts = [
    cfg.StrOpt(
        'file_system_dir',
        help='Directory to store function packages.'
    ),
    cfg.StrOpt(
        'provider',
        default='local',
        choices=['local', 'swift'],
        help='Storage provider for function code package.'
    ),
]

KUBERNETES_GROUP = 'kubernetes'
kubernetes_opts = [
    cfg.StrOpt(
        'namespace',
        default='qinling',
        help='Resources scope created by Qinling.'
    ),
    cfg.IntOpt(
        'replicas',
        default=3,
        help='Number of desired replicas in deployment.'
    ),
    cfg.StrOpt(
        'kube_host',
        default='http://127.0.0.1:8001',
        help='Kubernetes server address, e.g. you can start a proxy to the '
             'Kubernetes API server by using "kubectl proxy" command.'
    ),
    cfg.BoolOpt(
        'use_api_certificate',
        default=True,
        help='Whether to use client certificates to connect to the '
             'Kubernetes API server.'
    ),
    cfg.StrOpt(
        'ssl_ca_cert',
        default='/etc/qinling/pki/kubernetes/ca.crt',
        help='Path to the CA certificate for qinling to use to connect to '
             'the Kubernetes API server.'
    ),
    cfg.StrOpt(
        'cert_file',
        default='/etc/qinling/pki/kubernetes/qinling.crt',
        help='Path to the client certificate for qinling to use to '
             'connect to the Kubernetes API server.'
    ),
    cfg.StrOpt(
        'key_file',
        default='/etc/qinling/pki/kubernetes/qinling.key',
        help='Path to the client certificate key file for qinling to use to '
             'connect to the Kubernetes API server.'
    ),
    cfg.StrOpt(
        'log_devel',
        default='INFO',
        choices=['DEBUG', 'INFO', 'WARNING', 'ERROR'],
        help='Log level for kubernetes operations.'
    ),
    cfg.ListOpt(
        'trusted_cidrs',
        item_type=cfg.types.String(),
        default=[],
        help='List of CIDR that have access to the services in '
             'Kubernetes, e.g. trusted_cidrs=127.0.0.1/32,198.72.124.109/32. '
             'If it is empty list, the default value is the host IP address '
             'that the qinling-engine service is running on.'
    )
]

ETCD_GROUP = 'etcd'
etcd_opts = [
    cfg.StrOpt(
        'host',
        default='127.0.0.1',
        help='Etcd service host address.'
    ),
    cfg.PortOpt(
        'port',
        default=2379,
        help='Etcd service port.'
    ),
    cfg.StrOpt(
        'protocol',
        default='https',
        choices=['http', 'https'],
        help='Etcd connection protocol.'
    ),
    cfg.StrOpt(
        'ca_cert',
        default='/etc/qinling/pki/etcd/ca.crt',
        help='Path to CA certificate file to use to securely '
             'connect to etcd server.'
    ),
    cfg.StrOpt(
        'cert_file',
        default='/etc/qinling/pki/etcd/qinling-etcd-client.crt',
        help='Path to client certificate file to use to securely '
             'connect to etcd server.'
    ),
    cfg.StrOpt(
        'cert_key',
        default='/etc/qinling/pki/etcd/qinling-etcd-client.key',
        help='Path to client certificate key file to use to securely '
             'connect to etcd server.'
    ),
]

RLIMITS_GROUP = 'resource_limits'
rlimits_opts = [
    cfg.IntOpt(
        'default_cpu',
        default=100,
        help='Default cpu resource(unit: millicpu).'
    ),
    cfg.IntOpt(
        'min_cpu',
        default=100,
        help='Minimum cpu resource(unit: millicpu).'
    ),
    cfg.IntOpt(
        'max_cpu',
        default=300,
        help='Maximum cpu resource(unit: millicpu).'
    ),
    cfg.IntOpt(
        'default_memory',
        default=33554432,
        help='Default memory resource(unit: bytes).'
    ),
    cfg.IntOpt(
        'min_memory',
        default=33554432,
        help='Minimum memory resource(unit: bytes).'
    ),
    cfg.IntOpt(
        'max_memory',
        default=134217728,
        help='Maximum memory resource(unit: bytes).'
    ),
    cfg.IntOpt(
        'default_timeout',
        default=5,
        help='Default function execution timeout(unit: seconds)'
    ),
    cfg.IntOpt(
        'min_timeout',
        default=1,
        help='Minimum function execution timeout(unit: seconds).'
    ),
    cfg.IntOpt(
        'max_timeout',
        default=300,
        help='Maximum function execution timeout(unit: seconds).'
    ),
]


def list_opts():
    keystone_middleware_opts = auth_token.list_opts()
    keystone_loading_opts = [(
        'keystone_authtoken', loading.get_auth_plugin_conf_options('password')
    )]

    qinling_opts = [
        (API_GROUP, api_opts),
        (PECAN_GROUP, pecan_opts),
        (ENGINE_GROUP, engine_opts),
        (STORAGE_GROUP, storage_opts),
        (KUBERNETES_GROUP, kubernetes_opts),
        (ETCD_GROUP, etcd_opts),
        (RLIMITS_GROUP, rlimits_opts),
        (None, [launch_opt]),
        (None, default_opts),
    ]

    return keystone_middleware_opts + keystone_loading_opts + qinling_opts


def parse_args(args=None, usage=None, default_config_files=None):
    CLI_OPTS = [launch_opt]
    CONF.register_cli_opts(CLI_OPTS)

    for group, options in list_opts():
        CONF.register_opts(list(options), group)

    _DEFAULT_LOG_LEVELS = [
        'eventlet.wsgi.server=WARN',
        'oslo_service.periodic_task=INFO',
        'oslo_service.loopingcall=INFO',
        'oslo_db=WARN',
        'oslo_concurrency.lockutils=WARN',
        'kubernetes.client.rest=%s' % CONF.kubernetes.log_devel,
        'keystoneclient=INFO',
        'requests.packages.urllib3.connectionpool=CRITICAL',
        'urllib3.connectionpool=CRITICAL',
        'cotyledon=INFO',
        'futurist.periodics=WARN'
    ]
    default_log_levels = log.get_default_log_levels()
    default_log_levels.extend(_DEFAULT_LOG_LEVELS)
    log.set_defaults(default_log_levels=default_log_levels)
    log.register_options(CONF)

    CONF(
        args=args,
        project='qinling',
        version=version,
        usage=usage,
        default_config_files=default_config_files
    )
